cranelift/codegen/src/isa/x64/lower.isle

```diff
@@ -1452,6 +1452,20 @@
 (rule (lower (trap code))
       (side_effect (x64_ud2 code)))
 
+;;;; Rules for `trapif` ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+;; The flags must not have been clobbered by any other instruction between the
+;; iadd_ifcout and this instruction, as verified by the CLIF validator; so we
+;; can simply use the flags here.
+(rule (lower (trapif cc (iadd_ifcout _ _) tc))
```

Comment:
Hmm -- I'm not sure about this: we're matching that `iadd_ifcout` is our input, but we don't actually *use* the input, rather we implicitly depend on it clobbering flags. This then breaks our producer-consumer coupling as well, necessitating the `emit_consumer` helper, which feels pretty unsafe to me.

The most immediate issue is that if the `iadd_ifcout` isn't otherwise used, it won't actually be lowered, and then we're going to use bogus flags to do our trap. That seems like the biggest issue. I suspect this case simply isn't covered in our existing tests.

More broadly, we want to remove `flags` types competely eventually, because they're pretty error-prone (see #3249). But until then, we should work on doing something safer (at least correct, ideally more elegant as well!).

What about this: we write a constructor that takes the flags value (so in the pattern, capture it with `flags @ (iadd_ifcout _ _)`), and produces a `ProducesFlags.AlreadyExistingFlags` (that contains no actual inst), and handle that in `with_flags`. This ctor will need to call out to an external constructor to bump the refcount on the iadd_ifcout` as well (normally `put_in_regs` does this in an implicit converter from `Value` to `Reg`). Let's add a test where the `iadd_ifcout` is used *only* for its flags, and make sure it lowers properly.

There's also a relevant comment near the lowering rules for `iadd_ifcout` that should be updated I think.

Thoughts?